gpbft/chain.go

```diff
@@ -192,25 +149,29 @@ func (c ECChain) Validate() error {
 	if c.IsZero() {
 		return nil
 	}
-	var epochSoFar int64
 	for _, tipSet := range c {
-		switch {
-		case tipSet.IsZero():
+		if len(tipSet) == 0 {
 			return errors.New("chain cannot contain zero-valued tip sets")
-		case tipSet.Epoch <= epochSoFar:
-			return errors.New("chain epoch must be in order and unique")
-		default:
-			epochSoFar = tipSet.Epoch
 		}
 	}
 	return nil
 }
 
+// Returns an identifier for the chain suitable for use as a map key.
+// This must completely determine the sequence of tipsets in the chain.
+func (c ECChain) Key() ChainKey {
+	buf := bytes.Buffer{}
+	for _, t := range c {
+		buf.Write(t)
```

Comment:
Minor comment: since the length of writes is known, grow buffer beforehand.
```suggestion
	var buf bytes.Buffer
	for _, t := range c {
		buf.Grow(len(c))
```